### PR TITLE
Simple custom filters editor

### PR DESCRIPTION
**Pull request by chrmod**
fix https://github.com/ghostery/ghostery-extension/issues/22

A custom filter editor is a way for filter maintainers and users to apply and test their own adblocking rules.

Initial rollout will have it hidden in the "developer section". In near future this will become available to all users.

### CODE REVIEW

File: extension-manifest-v3/src/utils/custom-filters.js [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+ */
+
+const KEY_DNR_RULES = 'custom-filters-dnr';
+
+export async function getDNRRules() {
+  return (await chrome.storage.local.get([KEY_DNR_RULES]))[KEY_DNR_RULES] || [];
+}
+
+export async function setDNRRules(filters = []) {
+  await chrome.storage.local.set({
```
Comment (chrmod) [inline extension-manifest-v3/src/utils/custom-filters.js:19]: I'm not convinced if this storage is needed. If we don't want to show the DNR conversion output (after the reload) in UI we can remove it.

**Comment by smalluban** [comment]
Something is wrong specifically in Safari with page <-> BG communication on custom filters update - I'm trying to find what, that's why I have changed the PR to draft.

File: extension-manifest-v3/src/background/utils/setup.js [hunk 2]
```diff
@@ -14,7 +14,7 @@ export default function asyncSetup(promises, threshold = 5000) {
 
   const result = {
     pending: Promise.race([
-      Promise.all(promises),
```
Comment (chrmod) [inline extension-manifest-v3/src/background/utils/setup.js:17]: why is this needed?

File: extension-manifest-v3/src/pages/settings/components/custom-filters.js [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+    const results = await Promise.allSettled(
+      [...networkFilters].map((filter) => converter.convert(filter)),
+    );
+
+    for (const result of results) {
+      dnrErrors.push(...result.value.errors);
+      dnrRules.push(...result.value.rules);
+    }
+
+    // DNR validation is not required for Firefox, but it wont harm
```
Comment (chrmod) [inline extension-manifest-v3/src/pages/settings/components/custom-filters.js:62]: we can remove the comment

File: extension-manifest-v3/src/pages/settings/components/custom-filters.js [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+    });
+
+    host.dnrRules = dnrRules;
+  }
+
+  // Update engine
+  await chrome.runtime.sendMessage({
+    action: 'custom-filters:engine',
+    filters: [
+      ...(__PLATFORM__ === 'firefox' ? networkFilters : []),
```
Comment (chrmod) [inline extension-manifest-v3/src/pages/settings/components/custom-filters.js:80]: we have to pass network filters also for DNR platforms are there are certain network filter options that affect cosmetic filters `$elemhide, $ehide, $specifichide, $shide, $generichide, $ghide`

File: extension-manifest-v3/src/utils/engines.js [hunk 5]
```diff
@@ -15,10 +15,17 @@ import {
...
   getLinesWithFilters,
   mergeDiffs,
+  Config,
 } from '@cliqz/adblocker';
 
 import { registerDatabase } from '/utils/indexeddb.js';
 
+export const CUSTOM_ENGINE = 'custom-filters';
+
+function isCustomEngine(name) {
```
Comment (chrmod) [inline extension-manifest-v3/src/utils/engines.js:25]: do we need this function? or is it a leftover?

File: extension-manifest-v3/src/utils/engines.js [hunk 6]
```diff
@@ -29,6 +36,41 @@ function saveToMemory(name, engine) {
   engines.set(name, engine);
 }
 
+// custom filter exceptions should apply to all engines
+function shareExceptions(name, engine) {
+  if (name.startsWith('custom-filters')) {
```
Comment (chrmod) [inline extension-manifest-v3/src/utils/engines.js:41]: can check with `===`

File: extension-manifest-v3/src/background/utils/setup.js (hunk 2 above)
Comment (smalluban) [reply extension-manifest-v3/src/background/utils/setup.js]: I was in one of the implementation when I was passing ternary in `adblocker.js` - but it is not now, I'll clean it up.

**Comment by smalluban** [comment]
@chrmod PR ready to be re-reviewed ;)

**Review by smalluban — APPROVED** [review]

File: extension-manifest-v3/src/pages/settings/components/custom-filters.js [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+
+async function submitFilters(host) {
+  const { networkFilters, cosmeticFilters } = host.filters;
+
+  // Update DNR
+  if (__PLATFORM__ !== 'firefox') {
+    const dnrRules = [];
+    const dnrErrors = [];
+    const results = await Promise.allSettled(
+      [...networkFilters].map((filter) => converter.convert(filter)),
```
Comment (chrmod) [inline extension-manifest-v3/src/pages/settings/components/custom-filters.js:54]: no need for array spread

Comment (smalluban) [reply extension-manifest-v3/src/pages/settings/components/custom-filters.js:54]: It's a Set. Look above. I've changed that to simply remove duplicated filters (I discovered them while testing).

**Comment by Kees1958** [comment]
@chrmod I installed Ghostery today, but can't find the "Simple custom filters editor"  where can I test/try it?

**Comment by smalluban** [comment]
This is currently behind the developer tools flag. It means that it is in internal testing phase. 

However, if you want to check it out, click 5 times version of the extension on the home page of the settings.

**Comment by Kees1958** [comment]
@smalluban  Thanks for the fast response. 

URL filters seem to work well, so that is great. I also tried some simple cosmetic filters (e.g. hiding empty ad paceholders on Dutch news websites).

This is great thanks
